components/server/src/workspace/workspace-service.ts

```diff
@@ -355,11 +356,57 @@ export class WorkspaceService {
                     reason,
                     policy,
                 );
+                await this.updateDeletionEligabilityTime(userId, info.workspace.id);
             }),
         );
         return infos.map((instance) => instance.workspace);
     }
 
+    /**
+     * Sets the deletionEligibilityTime of the workspace, depening of the current state of the workspace and the configuration.
+     *
+     * @param userId sets the
+     * @param workspaceId
+     * @returns
+     */
+    async updateDeletionEligabilityTime(userId: string, workspaceId: string): Promise<void> {
+        try {
+            const daysToLive = this.config.workspaceGarbageCollection?.minAgeDays || 14;
+            const daysToLiveForPrebuilds = this.config.workspaceGarbageCollection?.minAgePrebuildDays || 7;
+
+            const workspace = await this.doGetWorkspace(userId, workspaceId);
+            const instance = await this.db.findCurrentInstance(workspaceId);
+            const lastActive =
+                instance?.stoppingTime || instance?.startedTime || instance?.creationTime || workspace?.creationTime;
+            if (!lastActive) {
+                return;
+            }
+            const deletionEligibilityTime = new Date(lastActive);
+            if (workspace.type === "prebuild") {
+                // set to last active plus daysToLiveForPrebuilds as iso string
+                deletionEligibilityTime.setDate(deletionEligibilityTime.getDate() + daysToLiveForPrebuilds);
+                await this.db.updatePartial(workspaceId, {
+                    deletionEligibilityTime: deletionEligibilityTime.toISOString(),
+                });
+                return;
+            }
+            deletionEligibilityTime.setDate(deletionEligibilityTime.getDate() + daysToLive);
+            // workspaces with pending changes live twice as long
+            if (
+                (instance?.gitStatus?.totalUncommitedFiles || 0) > 0 ||
+                (instance?.gitStatus?.totalUnpushedCommits || 0) > 0 ||
+                (instance?.gitStatus?.totalUnpushedCommits || 0) > 0
+            ) {
+                deletionEligibilityTime.setDate(deletionEligibilityTime.getDate() + daysToLive);
```

Comment:
nit: I had to check back to understand what's going on here, because the setDate/getDate combo called twice. Alternatively we could calculate + conditionally update and then set once.